### DISCUSSION TITLE
Does this library support writing jpg files with only 1 or 2 channels to a pdf?

### DISCUSSION THREAD

**Discussion by ChloeOfAstora** [AndreRenaud/PDFGen#132] [Q&A] [answered]
I am writing a simple file conversion program where I have a bunch of images in a folder and turn those into a pdf at some point, I just made a function to enable converting the images to greyscale only, and while the file size of the images is a lot lower, the pdf only shrinks a couple kb in size on a 42mb pdf.
Is this because there is no support for using images with fewer channels in this library (or pdf's in general) or am I just using the wrong function for this?
Right now i'm using the pdf_add_image_file function.

**Comment by AndreRenaud** [comment, marked as answer]
The JPEG images basically get copied in unchanged, since PDF supports JPEG decoding natively. So you should see a file size reduction. Can you send me an example of the problem (ie: the jpeg files etc...)?

**Reply by ChloeOfAstora** [reply]
here is one example image in both greyscale and colour, 
![000170 - bw](https://github.com/AndreRenaud/PDFGen/assets/138579561/77b2b8ff-1b2a-4c84-8745-70272a8ee3ee)
![000170 - colour](https://github.com/AndreRenaud/PDFGen/assets/138579561/1e65114a-1203-4043-95d0-d5df56d100d9)
![afbeelding](https://github.com/AndreRenaud/PDFGen/assets/138579561/afc4b599-fcc7-4188-a6ee-f73565ffaa90)

and here are the pdf sizes for the 2 versions with 300+ colour/grayscale images
![afbeelding](https://github.com/AndreRenaud/PDFGen/assets/138579561/d6b2f693-2ead-432e-9c72-61496f671b7f)

maybe I just messed up somewhere else in my code but I was expecting a bigger difference.

**Comment by AndreRenaud** [comment]
Those numbers don't seem too surprising to me - the greyscale version is only 9kB smaller (7%). The final PDF is ~6% smaller. So that roughly tracks. The surprising thing is that making the jpeg greyscale in the first place didn't save much space.

Using the images you sent through, I was able to get slightly better compression (16%) using ImageMagick - how are you converting them from colour -> grayscale?

I used: 
`convert colour.jpg -colorspace Gray grey.jpg`

The initial size of the image you sent was 130750, your grayscale one was 121602 (7% reduction), imagemagick's grayscale one is 110109 (16% reduction).

**Reply by ChloeOfAstora** [reply]
I've been converting them with this code:

void convertToGrayscale(unsigned char* data, int width, int height, int channels, unsigned char* gsdata) {
    for (int i = 0; i < width * height; ++i) {
        // Calculate grayscale value using the weighted method (luma)
        unsigned char gray = 0;

        if (channels == 1 || channels == 2) {
            // Grayscale image, no need for conversion
            gray = data[i];
            if (channels == 2) {
                // Grayscale image, no need for conversion
                gsdata[i] = data[i + 1];
            }
        } else if (channels >= 3) {
            // RGB or RGBA image
            unsigned char r = data[i * channels];
            unsigned char g = data[i * channels + 1];
            unsigned char b = data[i * channels + 2];

            gray = static_cast<unsigned char>(0.21f * r + 0.72f * g + 0.07f * b);

            if (channels == 4) {
                // Alpha channel exists, preserve it
                unsigned char alpha = data[i * channels + 3];
                gsdata[i + 1] = alpha;
            }
        }
        gsdata[i] = gray;

    }
}

I think one of the factors in making the difference so little might also be the jpg compression, both images were compressed at 65% quality, when set to 100 the difference in pdf sizes is a lot bigger 
![afbeelding](https://github.com/AndreRenaud/PDFGen/assets/138579561/a6981c47-717b-4377-ac19-f2e880bcb148)
edit - cant get the code block to work properly, sorry

**Reply by AndreRenaud** [reply]
Ah, I see. Yeah, I wouldn't recommend embedding them as raw RGB data - that will make the PDF absolutely massive. You want to embed them as JPEG files directly (unmodified).

Do you have your code somewhere that I could look at?

**Reply by ChloeOfAstora** [reply]
yes, my code is at https://github.com/ChloeOfAstora/cbz2pdf
I do save it to a jpg file with compression with stb_image_write.h

**Reply by AndreRenaud** [reply]
I suspect that `stbi_write_jpg` doesn't do as good of a job at compressing things as other JPEG libraries. It is designed for small code size/simplicity, but not bulk performance. Could you do the conversion outside your c++ code using imagmagick? I see you're basically unpacking a .zip file, and converting each one to grayscale, before deleting it. If you don't care about the speed of this, you could even just do `system("convert " + tempPath.string() + "/" + entry.path().filename().string() + " " + filenameNew)` (ignore the terrible string handling).

**Reply by ChloeOfAstora** [reply]
Yes I could do that, before this I was using a bash script that I wrote to do basically the same thing, just using a lot of external utilities like imagmagick, I wanted to try making a program that would just be one file and work everywhere. So if I want to keep it internal in the program I guess I'll have to look for a better performance jpg library.

**Reply by AndreRenaud** [reply]
Yeah, that makes sense. Having it all nicely self-contained is a good feature (I mean, that's the whole point of this PDF library).
Unfortunately there's not much that PDFGen can do in this case - apart from parsing the header to work out the width/height, the reset of the JPEG data is just copied unchanged into the PDF.

**Reply by ChloeOfAstora** [reply]
Thank you very much for your help and fast responses!
And also thank you very much for your amazing work with this library.

**Comment by AndreRenaud** [comment]
Thanks. I'm glad you're getting some use out of it. Good luck with the jpeg compression stuff.
